src/eddymotion/model.py

```diff
@@ -271,6 +270,102 @@ def predict(self, gradient, **kwargs):
         return self._data
 
 
+class PETModel:
+    """A PET imaging realignment model based on B-Spline approximation."""
+
+    __slots__ = ("_t", "_x", "_order", "_coeff", "_mask", "_shape", "_n_ctrl")
+
+    def __init__(self, timepoints=None, n_ctrl=None, mask=None, order=3, **kwargs):
+        """
+        Create the B-Spline interpolating matrix.
+
+        Parameters:
+        -----------
+        timepoints : :obj:`list`
+            The timing (in sec) of each PET volume.
+            E.g., ``[20.,   40.,   60.,  120.,  180.,  240.,  360.,  480.,  600.,
+            900., 1200., 1800., 2400., 3000.]``
+
+        n_ctrl : :obj:`int`
+            Number of B-Spline control points. If `None`, then one control point every
+            six timepoints will be used. The less control points, the smoother is the
+            model.
+
+        """
+        if timepoints is None:
+            raise TypeError("timepoints must be provided in initialization")
+
+        self._order = order
+        self._mask = mask
+
+        self._x = np.array(timepoints, dtype="float32")
+
+        # Calculate index coordinates in the B-Spline grid
+        self._n_ctrl = n_ctrl or (len(timepoints) // 4) + 1
+
+        # B-Spline knots
+        self._t = np.arange(-3, float(self._n_ctrl) + 4, dtype="float32")
+
+    def fit(self, data, *args, **kwargs):
+        """Do nothing."""
+        from scipy.interpolate import BSpline
+        from scipy.sparse.linalg import cg
+
+        n_jobs = kwargs.pop("n_jobs", None) or 1
+
+        timepoints = kwargs.get("timepoints", None) or self._x
+        x = (
+            (np.array(timepoints, dtype="float32") - self._x[0]) / self._x[-1]
+        ) * self._n_ctrl
+
+        self._shape = data.shape[:3]
+
+        # Convert data into V (voxels) x T (timepoints)
+        data = (
+            data.reshape((-1, data.shape[-1]))
+            if self._mask is None else data[self._mask]
+        )
+
+        # A.shape = (T, K - 4); T= n. timepoints, K= n. knots (with padding)
+        A = BSpline.design_matrix(x, self._t, k=self._order)
+        AT = A.T
+        ATdotA = AT @ A
+
+        # One single CPU - linear execution (full model)
+        if n_jobs == 1:
+            self._coeff = np.array([
+                cg(ATdotA, AT @ v)[0] for v in data
+            ])
+            return
+
+        # Parallelize process with joblib
+        with Parallel(n_jobs=n_jobs) as executor:
+            results = executor(
+                delayed(cg)(ATdotA, AT @ v)
+                for v in data
+            )
+
+        self._coeff = np.array([r[0] for r in results])
+
+    def predict(self, timepoint, **kwargs):
```

Comment:
Random thought here but also relevant to the other models: do we want to raise an error if `predict` is called without first calling `fit`? Potentially store a `.is_fitted_` attribute and check for it, a la sklearn estimators.

Reply:
+1

Reply:
Opened #113